workflow/controller/dag.go

```diff
@@ -87,7 +87,29 @@ func (d *dagContext) assessDAGPhase(targetTasks []string, nodes map[string]wfv1.
 			retriesExhausted = false
 		}
 	}
+
 	if unsuccessfulPhase != "" {
+		// If failFast set to false, we should return Running to continue this workflow for other DAG branch
+		if d.wf.Spec.FailFast != nil && !*d.wf.Spec.FailFast {
+			tmpOverAllFinished := true
+			// If all the nodes have finished, we should mark the failed node to finish overall workflow
+			// So we should check all the targetTasks have finished
+			for _, tmpDepName := range targetTasks {
+				tmpDepNode := d.getTaskNode(tmpDepName)
+				if tmpDepNode == nil {
+					tmpOverAllFinished = false
```

Comment:
I think there may be a bug here: if FailFast is disabled, and a branch fails, then we will always reach this if-block and the DAG will never complete.

Reply:
Not yet.   

If a branch failed,  the `unsuccessfulPhase` isn't empty, so the `if condition` will get in.   If at the same time, there's still other branch running, then `tmpOverAllFinished` will be set to `false`, the `assessDAGPhase` function will return `Running` to continue.

If all the branch has finished.  in the `for _, tmpDepName := range targetTasks ` function will find that all node has finished including `retryType` node., `tmpOverAllFinished` will be set `true`, so 

```
if retriesExhausted {
    return unsuccessfulPhase
}
```

function will be trigger, the `assessDAGPhase` function will return `NodeFailed`, so all of the workflow will be marked `Failed`.  This is in line with expectations.  

Our test case, `test/e2e/functional/dag-disable-failFast.yaml` will show the demo, it will finished when all the branch complete, but one of the branch is failed.

Reply:
Thanks for explanation. Yes you are right.